Playwright 🎭  🛠️

TanayParikh opened the pull request:
Enable Playwright Tests

Fixes: https://github.com/dotnet/aspnetcore/issues/30761
Fixes: https://github.com/dotnet/aspnetcore/issues/30882

TanayParikh committed:
Version Bumps

TanayParikh committed:
Run with `Test: Windows Server x64`

TanayParikh committed:
Install Playwright in CI

TanayParikh committed:
Add running playwright locally instructions

TanayParikh committed:
Test Cleanup

TanayParikh committed:
Remove Selenium

TanayParikh committed:
Prebuild for Playwright install

TanayParikh committed:
Update default-build.yml

TanayParikh committed:
Update default-build.yml

TanayParikh committed:
Update default-build.yml

TanayParikh committed:
Update default-build.yml

TanayParikh committed:
Update default-build.yml

TanayParikh committed:
Playwright install in test

mkArtakMSFT commented on src/ProjectTemplates/README.md:
```diff
@@ -71,6 +71,14 @@ Then, use one of:
 previous step, it is NOT advised that you install templates created on your local machine using just
 `dotnet new -i [nupkgPath]`.
 
+#### Running Playwright tests
```
Based on the title it seems like this is for running all playwright tests, but it's referring to only the `Blazor Template Tests`. Should the title be updated?

TanayParikh committed:
New template test pipeline

TanayParikh committed:
Update components-e2e-tests.yml

TanayParikh committed:
Update components-template-tests.yml

TanayParikh committed:
Update components-e2e-tests.yml

TanayParikh committed:
Update components-e2e-tests.yml

TanayParikh committed:
Update components-e2e-tests.yml

TanayParikh committed:
Update components-template-tests.yml

TanayParikh committed:
Update components-e2e-tests.yml

TanayParikh committed:
Update components-e2e-tests.yml

TanayParikh committed:
Update components-e2e-tests.yml

TanayParikh committed:
Revert new pipeline changes

TanayParikh committed:
RestoreAdditionalProjectSources

TanayParikh committed:
Update Templates.Blazor.Tests.csproj

TanayParikh committed:
PR Feedback

TanayParikh committed:
RunBlazorPlaywrightTemplateTests

TanayParikh committed:
Socket logic

TanayParikh committed:
Fixed socket wait logic

Fixes https://github.com/dotnet/aspnetcore/issues/30761

TanayParikh committed:
Remove redundant process wait for exit

We're already in the `Exited` Event:

There are two ways of being notified when the associated process exits: synchronously and asynchronously. Synchronous notification means calling the WaitForExit method to block the current thread until the process exits. Asynchronous notification uses the Exited event, which allows the calling thread to continue execution in the meantime. In the latter case, EnableRaisingEvents must be set to true for the calling application to receive the Exited event.

https://learn.microsoft.com/dotnet/api/system.diagnostics.process.exited?view=net-7.0

TanayParikh commented on src/Shared/Process/ProcessEx.cs:
```diff
@@ -179,7 +179,7 @@ private void OnProcessExited(object sender, EventArgs e)
             }
         }
-        _process.WaitForExit();
+
```
We're already in the `Exited` Event so this call is redundant:

> There are two ways of being notified when the associated process exits: synchronously and asynchronously. Synchronous notification means calling the WaitForExit method to block the current thread until the process exits. Asynchronous notification uses the Exited event, which allows the calling thread to continue execution in the meantime. In the latter case, EnableRaisingEvents must be set to true for the calling application to receive the Exited event.

https://learn.microsoft.com/dotnet/api/system.diagnostics.process.exited?view=net-7.0

BrennanConroy replied:
Pretty sure this is still needed when using async IO. It guarantees that stdout and stderr are done.
Data point; after this PR was merged there are now failures like:
```
2023-02-14T17:40:24.0586118Z [xUnit.net 00:00:04.23]         | [3.247s] Templates.Blazor.Test.BlazorWasmTemplateTest Information: Process exited.
2023-02-14T17:40:24.0586151Z [xUnit.net 00:00:04.23]         | [3.248s] Templates.Blazor.Test.BlazorWasmTemplateTest Information: [ERROR] No templates or subcommands found matching: 'reactredux'.
2023-02-14T17:40:24.0586182Z [xUnit.net 00:00:04.23]         | [3.248s] Templates.Blazor.Test.BlazorWasmTemplateTest Information: [ERROR] 
2023-02-14T17:40:24.0586210Z [xUnit.net 00:00:04.23]         | [3.248s] Templates.Blazor.Test.BlazorWasmTemplateTest Information: [ERROR] To list installed templates similar to 'reactredux', run:
2023-02-14T17:40:24.0594633Z [xUnit.net 00:00:04.23]         | [3.248s] Templates.Blazor.Test.BlazorWasmTemplateTest Information: [ERROR]    dotnet new list reactredux
2023-02-14T17:40:24.0627195Z [xUnit.net 00:00:04.23]         | [3.248s] Templates.Blazor.Test.BlazorWasmTemplateTest Information: [ERROR] To search for the templates on NuGet.org, run:
2023-02-14T17:40:24.0629366Z [xUnit.net 00:00:04.23]         | [3.248s] Templates.Blazor.Test.BlazorWasmTemplateTest Information: [ERROR]    dotnet new search reactredux
2023-02-14T17:40:24.0736529Z [xUnit.net 00:00:04.23]         | [3.248s] Templates.Blazor.Test.BlazorWasmTemplateTest Information: [ERROR] 
2023-02-14T17:40:24.0754965Z [xUnit.net 00:00:04.23]         | [3.248s] Templates.Blazor.Test.BlazorWasmTemplateTest Information: [ERROR] 
2023-02-14T17:40:24.0774833Z [xUnit.net 00:00:04.23]         | [3.248s] Templates.Blazor.Test.BlazorWasmTemplateTest Information: [ERROR] For details on the exit code, refer to https://aka.ms/templating-exit-codes#103
2023-02-14T17:40:24.0794693Z [xUnit.net 00:00:04.23]         | [3.254s] Templates.Blazor.Test.BlazorWasmTemplateTest Error: Test threw an exception.
2023-02-14T17:40:24.0812567Z [xUnit.net 00:00:04.23]         | System.InvalidOperationException: Failed to uninstall previous templates. The template 'reactredux' could still be found.
2023-02-14T17:40:24.0901861Z [xUnit.net 00:00:04.23]         |    at Templates.Test.Helpers.TemplatePackageInstaller.VerifyCannotFindTemplateAsync(ITestOutputHelper output, String templateName) in /_/src/ProjectTemplates/Shared/TemplatePackageInstaller.cs:line 151
2023-02-14T17:40:24.0920908Z [xUnit.net 00:00:04.23]         |    at Templates.Test.Helpers.TemplatePackageInstaller.InstallTemplatePackages(ITestOutputHelper output) in /_/src/ProjectTemplates/Shared/TemplatePackageInstaller.cs:line 114
```
You can see from the test output that the text "No templates or subcommands found matching:" was output, however the test threw on the check for that exact string:
https://github.com/dotnet/aspnetcore/blob/fb1e6a74373f1e4f09cc1b0adaac5f0535e5f07d/src/ProjectTemplates/Shared/TemplatePackageInstaller.cs#L149-L151
This implies a race condition where the process exits and we grab the output before the stdout/stderr completed writing.
cc @wtgodbe

wtgodbe replied:
Thanks for catching this, I'll put up a PR to revert this line

TanayParikh committed:
Update BlazorWasmTemplateTest.cs

TanayParikh commented:
/azp run

TanayParikh committed:
Update Templates.Blazor.Tests.csproj

TanayParikh committed:
Revert App.Ref, add src in props

TanayParikh committed:
Config updates

TanayParikh committed:
Update RequiresDelayedBuildProjects.props

TanayParikh committed:
Update Templates.Blazor.Tests.csproj

TanayParikh committed:
Merge branch 'main' into taparik/playwright

TanayParikh committed:
Skip failing tests

TanayParikh commented on src/ProjectTemplates/test/Templates.Blazor.Tests/BlazorWasmTemplateTest.cs:
```diff
@@ -59,7 +59,7 @@ private static async Task<IPage> NavigateToPage(IBrowserContext browser, string
     }
 
-    [Theory(Skip = "https://github.com/dotnet/aspnetcore/issues/30882")]
+    [Theory(Skip="Unblock CI")]
```
Will replace with issue & tackle separately.

TanayParikh replied:
CI Issues tracked in https://github.com/dotnet/aspnetcore/issues/46430

mkArtakMSFT commented on .azure/pipelines/ci.yml:
```diff
@@ -697,7 +697,7 @@ stages:
           cancelTimeoutInMinutes: 30
           buildArgs: -all -pack -test -binaryLog /p:SkipHelixReadyTests=true /p:SkipIISNewHandlerTests=true /p:SkipIISTests=true
-                     /p:SkipIISExpressTests=true /p:SkipIISNewShimTests=true /p:RunTemplateTests=false
+                     /p:SkipIISExpressTests=true /p:SkipIISNewShimTests=true /p:RunTemplateTests=false /p:RunBlazorPlaywrightTemplateTests=true
```
Will setting this new variable (RunBlazorPlaywrightTemplateTests) to `false` result in the current behavior (as if we don't have this change in place)?

TanayParikh replied:
It's fine to just omit `RunBlazorPlaywrightTemplateTests` in that case as it'll default to using the value of `RunTemplateTests`, but yes setting it to `false` would work as well.

https://github.com/dotnet/aspnetcore/blob/20ebf882bf561130d3288a419ee2caa6f3d343c4/src/ProjectTemplates/test/Templates.Blazor.Tests/Templates.Blazor.Tests.csproj#L10-L11

mkArtakMSFT commented on src/ProjectTemplates/test/Templates.Blazor.Tests/BlazorWasmTemplateTest.cs:
```diff
@@ -107,7 +107,7 @@ private static async Task AssertCompressionFormat(AspNetProcess aspNetProcess, s
     }
 
-    [Theory(Skip = "https://github.com/dotnet/aspnetcore/issues/30882")]
+    [Theory(Skip = "https://github.com/dotnet/aspnetcore/issues/45736")]
```
The issue here was blocked on a Playwright issue, which they have resolved with the following comment: 
<img width="690" alt="image" src="https://user-images.githubusercontent.com/34246760/216194844-361796de-0ed5-42ab-81e4-67e4f2d2fa66.png">

Does that impact our decisions regarding this in any way?
I assume you're trying to wrap up your work already, and leaving this to be handled as part of the work on the linked issue is the right thing to do. So mostly wanted to note the progress on the dependency.

wtgodbe replied:
Going to try un-skipping this

wtgodbe replied:
Nope, still failing

TanayParikh committed:
Add https://github.com/dotnet/aspnetcore/issues/46430

wtgodbe commented:
The problem is that the `Microsoft.AspNetCore.Components.WebAssembly` package hasn't been built yet at the time the tests run (other packages like `Microsoft.AspNetCore.Components.WebAssembly.Server` _are_ available). Trying to figure out if that always would have been the case or if something's changed here.

wtgodbe committed:
Update Templates.Blazor.Tests.csproj

wtgodbe reviewed (approved):
LGTM

wtgodbe committed:
Update BlazorWasmTemplateTest.cs

wtgodbe committed:
Update BlazorWasmTemplateTest.cs
